src/data/matrix/basic.lean

```diff
@@ -1044,6 +1044,30 @@ lemma transpose_reindex (eₘ : m ≃ l) (eₙ : n ≃ o) (M : matrix m n α) :
   (reindex eₘ eₙ M)ᵀ = (reindex eₙ eₘ Mᵀ) :=
 rfl
 
+variables {m₁ n₁ o₁ : Type*} [fintype m₁] [fintype n₁] [fintype o₁]
+/-- equivalences between matrix ((m × n) × o) ((m₁ × n₁) × o₁) R and
+ matrix (m × n × o) (m₁ × n₁ × o₁) R -/
+def index_assoc : matrix ((m × n) × o) ((m₁ × n₁) × o₁) R ≃ matrix (m × n × o) (m₁ × n₁ × o₁) R :=
+{ to_fun := λ A, reindex (equiv.prod_assoc _ _ _) (equiv.prod_assoc _ _ _) A,
+  inv_fun := λ A, reindex (equiv.prod_assoc _ _ _).symm (equiv.prod_assoc _ _ _).symm A,
+  left_inv := λ _, by simp only [equiv.symm_symm, reindex_apply, minor_minor, minor_id_id,
+    equiv.symm_comp_self],
+  right_inv := λ _, by simp only [equiv.symm_symm, reindex_apply, minor_minor, minor_id_id,
+    equiv.self_comp_symm],
+  }
+
+def linear_equiv_index_assoc [comm_semiring α] [add_comm_monoid R] [module α R] :
+  matrix ((m × n) × o) ((m₁ × n₁) × o₁) R ≃ₗ[α] matrix (m × n × o) (m₁ × n₁ × o₁) R :=
+{ to_fun := λ A, reindex (equiv.prod_assoc _ _ _) (equiv.prod_assoc _ _ _) A,
+  map_add' := λ _ _, by simp only [reindex_apply, minor_add, pi.add_apply],
+  map_smul' := λ _ _, by simp only [reindex_apply, minor_smul, pi.smul_apply],
+  inv_fun := λ A, reindex (equiv.prod_assoc _ _ _).symm (equiv.prod_assoc _ _ _).symm A,
+  left_inv := λ _, by simp only [equiv.symm_symm, reindex_apply, minor_minor, minor_id_id,
+    equiv.symm_comp_self],
+  right_inv := λ _, by simp only [equiv.symm_symm, reindex_apply, minor_minor, minor_id_id,
+    equiv.self_comp_symm],
+  }
```

Comment:
This one is just
```suggestion
matrix.reindex_linear_equiv (equiv.prod_assoc _ _ _) (equiv.prod_assoc _ _ _)
```
although `reindex_linear_equiv` is defined in `linear_algebra.matrix.reindex` so can't be used here.